pico/processor/src/main/java/io/helidon/pico/processor/ServiceAnnotationProcessor.java

```diff
@@ -0,0 +1,217 @@
+                } else {
+                    String thisModuleName = getThisModuleName(null);
+                    if (thisModuleName == null) {
+                        servicesToProcess().clearModuleName();
+                    } else {
+                        servicesToProcess().moduleName(thisModuleName);
+                    }
+                }
+                if (typeSuffix.get() != null) {
+                    servicesToProcess().lastKnownTypeSuffix(typeSuffix.get());
+                }
+                if (srcPath.get() != null) {
+                    servicesToProcess().lastKnownSourcePathBeingProcessed(srcPath.get().toPath());
+                }
+                if (moduleInfoFile.get() != null) {
+                    servicesToProcess().lastKnownModuleInfoFilePath(moduleInfoFile.get().toPath());
+                }
+            }
+
+            return super.process(annotations, roundEnv);
+        } catch (Throwable t) {
+            error(getClass().getSimpleName() + " error during processing; " + t
+                          + " @ " + Utils.rootStackTraceElementOf(t), t);
+            // we typically will not even get to this next line since the messager.error() call will trigger things to halt
+            throw new ToolsException("error during processing: " + t
+                                             + " @ " + Utils.rootStackTraceElementOf(t), t);
+        } finally {
+            if (roundEnv.processingOver()) {
+                servicesToProcess().clearModuleName();
+            }
+        }
+    }
+
+    String getThisModuleName(ModuleInfoDescriptor descriptor) {
+        // user pass it in? if so then that will short-circuit the name calculation part
+        String moduleName = Options.getOption(Options.TAG_MODULE_NAME).orElse(null);
+        if (moduleName != null) {
+            return moduleName;
+        }
+        return (descriptor == null) ? null : descriptor.name();
+    }
+
+    ModuleInfoDescriptor getThisModuleDescriptor(AtomicReference<String> typeSuffix,
+                                                 AtomicReference<File> moduleInfoFile,
+                                                 AtomicReference<File> srcPath) {
+        ModuleInfoDescriptor descriptor = tryFindModuleInfo(StandardLocation.SOURCE_OUTPUT,
+                                                              typeSuffix,
+                                                              moduleInfoFile,
+                                                              srcPath);
+        if (descriptor == null) {
+            descriptor = tryFindModuleInfo(StandardLocation.SOURCE_PATH, typeSuffix, moduleInfoFile, srcPath);
+            if (descriptor == null) {
+                // attempt to retrieve from src/main/java if we can't recover to this point
+                descriptor = getThisModuleDescriptorFromSourceMain(moduleInfoFile, srcPath);
+            }
+        }
```

Comment:
Use optional pls
```java
return tryFindModuleInfo(StandardLocation.SOURCE_OUTPUT, typeSuffix, moduleInfoFile, srcPath)
        .or(() -> tryFindModuleInfo(StandardLocation.SOURCE_PATH, typeSuffix, moduleInfoFile, srcPath))
        // attempt to retrieve from src/main/java if we can't recover to this point
        .or(() -> getThisModuleDescriptorFromSourceMain(moduleInfoFile, srcPath))
        ...

```

Reply:
Not sure that using optionals to replace control-flow statements that are part of the language is an improvement TBH.

Reply:
ur right it's a more a personal preference